### PR TITLE
[ADD] Added ThingsChangedMessage producing for post requests

### LINKED ISSUE #327: [Add] RabbitMQ to send messages to any listener regarding updates to the data in a CDP4-COMET server
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the COMET Web Services
- [x] I have searched [open](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
Add RaabitMQ to notify a listener of changes to any data in the CDP4-COMET server
  - [ ] Add RabbitMQ to message listeners that a change has occured on any kind of Thing. An IThingMessageProducer service needs to be implemented that handles messaging and needs to be injected into the SiteDirectory and EngineeringModel Carter Module POST message handling.
  - [ ] add RabbitMQ connection setttings to appsettings.json and include an option to enable/disable messaging
  - [ ]  Add RabbitMQ to docker-compose for easy deployment
  - [ ] Add a messaging library to the CDP4-COMET-SDK that handles everything around messaging with RabbitMQ, such as the IThingMessageProducer and IThingMessageConsumer services and the defintiion of the message payloads etc.
  - [ ] any thing update event, corresponding to a transaction, can contain multiple things that are created updated and deleted. The following data shall be part of an update message. ChangedThings, TimeStamp, AcorID. It must be noted that the 10-25 protocol does not make it explicit that a Thing has been deleted. This is computed by computing the difference between state n and n+1 of the container Thing that contains the Thing that has been deleted

> NOTE: make sure that any calls to the queue are non-blocking and that the CDP4-COMET server is impacted as less as possible (we don't want this to slow us down)

I suggest to use the following naming conventions:
  - Exchange name: `cdp4cometexchange` (or cdp4-comet-exchange or cdp4_comet_exchange)
  - Queue name: `thingqueue` (or thing-queue, or thing_queue)
  - The `thingqueue` queue shall be bound to the `cdp4cometexchange`  exchange

### Steps to Reproduce
NA

### System Configuration

- COMET Web Services version: 8.0.0-RC21

### PR DESCRIPTION
**Pull request by nathanatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WebServices-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)
closes #327 
### Description

 Add RaabitMQ to notify a listener of changes to any data in the CDP4-COMET server
  - [x] Add RabbitMQ to message listeners that a change has occured on any kind of Thing. An IThingMessageProducer service needs to be implemented that handles messaging and needs to be injected into the SiteDirectory and EngineeringModel Carter Module POST message handling.
  - [x] add RabbitMQ connection setttings to appsettings.json and include an option to enable/disable messaging
  - [x]  Add RabbitMQ to docker-compose for easy deployment
  - [x] Add a messaging library to the CDP4-COMET-SDK that handles everything around messaging with RabbitMQ, such as the IThingMessageProducer and IThingMessageConsumer services and the defintiion of the message payloads etc.
  - [x] any thing update event, corresponding to a transaction, can contain multiple things that are created updated and deleted. The following data shall be part of an update message. ChangedThings, TimeStamp, AcorID. It must be noted that the 10-25 protocol does not make it explicit that a Thing has been deleted. This is computed by computing the difference between state n and n+1 of the container Thing that contains the Thing that has been deleted

> NOTE: make sure that any calls to the queue are non-blocking and that the CDP4-COMET server is impacted as less as possible (we don't want this to slow us down)
  - [x] Implemented a resolvable background service to queue message to push

### CODE REVIEW

File: CometServer/Configuration/ServiceMessagingConfig.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    /// </summary>
+    public class ServiceMessagingConfig
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ServiceMessagingConfig"/> class.
+        /// </summary>
+        public ServiceMessagingConfig()
+        {
+            // set defaults
+            this.IsEnabled = true;
```
Comment (lxatrhea) [inline CometServer/Configuration/ServiceMessagingConfig.cs:40]: I don't think that hard coding stuff here is what we want. Either we have a settings file, or we don;t and then it should just does not work (and log why it doesn't work).

**Review by lxatrhea — CHANGES_REQUESTED** [review]

File: CometServer.Tests/Configuration/AppConfigServiceTestFixture.cs [hunk 2]
```diff
@@ -69,6 +69,11 @@ public void Verify_that_configuration_is_loaded_from_appsettings()
             Assert.That(appConfigService.AppConfig.HealthConfig.RequireHost, Is.Empty);
 
             Assert.That(appConfigService.AppConfig.Midtier.BacktierWaitTime, Is.EqualTo(333));
+
```
Comment (antoineatrhea) [inline CometServer.Tests/Configuration/AppConfigServiceTestFixture.cs:72]: Would be nice to wrap all Assert.That by a Assert.Multiple

File: CometServer/Configuration/ServiceMessagingConfig.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ServiceMessagingConfig"/> class.
+        /// </summary>
+        /// <param name="configuration">
+        /// The <see cref="IConfiguration"/> used to set the properties
+        /// </param>
+        public ServiceMessagingConfig(IConfiguration configuration)
+        {
+            this.IsEnabled = !bool.TryParse(configuration["MessageBroker:IsEnabled"], out var isEnabled) || isEnabled;
```
Comment (antoineatrhea) [inline CometServer/Configuration/ServiceMessagingConfig.cs:58]: you can directly use configuration.GetValue("...", defaultValue:...)

File: CometServer/Configuration/ServiceMessagingConfig.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+            this.UserName = configuration["MessageBroker:UserName"];
+            this.Password = configuration["MessageBroker:Password"];
+            this.HostName = configuration["MessageBroker:HostName"];
+        }
+
+        /// <summary>
+        /// Gets or sets the host name of the back tier.
+        /// </summary>
+        public string HostName { get; set; }
```
Comment (antoineatrhea) [inline CometServer/Configuration/ServiceMessagingConfig.cs:78]: Are we sure that would let all properties as public set ? should not be private set?

**Review by lxatrhea — APPROVED** [review]

**Review by antoineatrhea — APPROVED** [review]

**Pull request merged**